File: app/src/main/java/com/sdex/activityrunner/ApplicationListViewModel.java
package com.sdex.activityrunner;

import android.app.Application;
import android.arch.lifecycle.AndroidViewModel;
import android.arch.lifecycle.LiveData;
import android.arch.persistence.db.SimpleSQLiteQuery;
import android.content.SharedPreferences;
import android.preference.PreferenceManager;
import android.support.annotation.NonNull;

import com.sdex.activityrunner.db.AppDatabase;
import com.sdex.activityrunner.db.application.ItemModel;
import com.sdex.activityrunner.db.query.GetApplicationsQuery;
import com.sdex.activityrunner.preferences.SortingPreferences;

import java.util.List;

public class ApplicationListViewModel extends AndroidViewModel {

  private final AppDatabase appDatabase;
  private final SortingPreferences sortingPreferences;

  public ApplicationListViewModel(@NonNull Application application) {
    super(application);
    appDatabase = AppDatabase.getDatabase(application);
    SharedPreferences sharedPreferences =
      PreferenceManager.getDefaultSharedPreferences(application);
    sortingPreferences = new SortingPreferences(sharedPreferences);
  }

  public LiveData<List<ItemModel>> getItems(String searchText) {
    GetApplicationsQuery query =
      new GetApplicationsQuery(searchText, sortingPreferences);
    return appDatabase.getApplicationModelDao()
      .getApplicationModels(new SimpleSQLiteQuery(query.toString()));
  }

  public LiveData<List<ItemModel>> getItems() {
    return getItems(null);
  }
}

File: app/src/main/java/com/sdex/activityrunner/ApplicationListViewModel.kt
package com.sdex.activityrunner

import android.app.Application
import android.arch.lifecycle.AndroidViewModel
import android.arch.lifecycle.LiveData
import android.arch.persistence.db.SimpleSQLiteQuery
import android.preference.PreferenceManager
import com.sdex.activityrunner.db.AppDatabase
import com.sdex.activityrunner.db.application.ApplicationModel
import com.sdex.activityrunner.db.query.GetApplicationsQuery
import com.sdex.activityrunner.preferences.SortingPreferences

class ApplicationListViewModel(application: Application) : AndroidViewModel(application) {

  private val appDatabase: AppDatabase = AppDatabase.getDatabase(application)
  private val sortingPreferences: SortingPreferences

  init {
    val sharedPreferences = PreferenceManager.getDefaultSharedPreferences(application)
    sortingPreferences = SortingPreferences(sharedPreferences)
  }

  fun getItems(searchText: String?): LiveData<List<ApplicationModel>> {
    val query = GetApplicationsQuery(searchText, sortingPreferences)
    return appDatabase.applicationModelDao
      .getApplicationModels(SimpleSQLiteQuery(query.toString()))
  }
}
